### ISSUE TITLE
`publish` should continue if git tag exists

### ISSUE THREAD

**Issue by WoodyWoodsta** [lerna/lerna#1894]
I have a mono-repo of 20+ modules. My private, self-hosted registry supports publishing the same version more than once (poor practice, I know, but that's not the point). If I run `lerna publish`, and one of the resulting git tags already exist, then the publish process fails at some unknown point in the process, and most of my modules are left unpublished (but bumped in the `package.json` files).

## Expected Behavior
If tags are unsafe to create/push, then this should be the last step (specifically, after actually publishing the modules). If it's possible to make the operation of creating and pushing a git tag safer, then this can be done, so at least the rest of the modules can be published to reflect their new version on the filesystem.

## Current Behavior
Described in the opening paragraph

<details><summary>lerna.json</summary><p>

```json
{
  "packages": [
    "deployment/*",
    "packages/pdf-*",
    "tools/packages/*"
  ],
  "version": "independent",
  "npmClient": "yarn"
}

```
</p></details>

<details><summary>lerna-debug.log</summary><p>

```txt
199 error Error: Command failed: git tag xxx@1.0.3 -m xxx@1.0.3
199 error fatal: tag 'xxx@1.0.3' already exists
199 error
199 error     at makeError (/xxx/node_modules/execa/index.js:174:9)
199 error     at Promise.all.then.arr (/xxx/node_modules/execa/index.js:278:16)

```
</p></details>

## Your Environment

| Executable | Version |
| ---: | :--- |
| `lerna --version` | 3.10.7 |
| `npm --version`  | 6.4.1 |
| `yarn --version` | 1.13.1 |
| `node --version` | 10.15.0 |

| OS | Version |
| --- | --- |
| NAME | VERSION |
| macOS Mojave | 10.14.3 beta |

**Comment by WoodyWoodsta** [comment]
The _exact_ commmand was `lerna publish` and the _full_ error log (with sensitive details redacted) is as above.

Anyway, as someone who is not 100% clued up on the internals of `lerna`, a brief think about it makes me believe that this should happen whenever `lerna` attempts to create git tags, and so would include any `version` or `publish` commands without the `--no-git-tag-version` flag.

Hope this helps?

**Comment by evocateur** [comment]
I can agree that the duplicate tag error needs more graceful handling, but I don't agree that lerna should _allow_ duplicate version tags to be created/consumed (even implicitly). It's fundamentally conflicting with `lerna publish`'s basic operative contract: "find all the packages changed since the last tagged release and publish them to a registry".

Real Soon Now™ I'm going to completely sever `lerna publish` and `lerna version` (in a breaking change) so we can get some clarity about the responsibility of each command. One current factor muddying this discussion is that `lerna publish`, without any other arguments, will call `lerna version` internally (which is the actual command that attempts the `git` operations).

`lerna publish` in and of itself (that is, _without_ the implicit `lerna version`) should certainly be able to publish as many times as you want, according to the responses of your configured registry. Currently, however, this is not the case.

The reason I asked for the complete log is that there is a _lot_ more debug information that could potentially be useful.

**Comment by WoodyWoodsta** [comment]
I agree and I think that synchronises well with the idea of not ever publishing the same version of a module to NPM or any other registry more than once. Issues arise, however, when an NPM publish fails but the tags exist, leaving the local repo in a misrepresentative state (I know that this is already covered in a separate set of issues). So either the git operations are made less rigid, or they are performed after a confirmed, successful publish, or change detection is done locally as well as against the registry (in which case the above issue could have been solved by running `lerna publish` again, after correcting registry issues, bringing the registry in line with the already created tags).

Glad to here that there is work being done around this and if the decoupling of `publish` and `version` is something coming in the near future, and which would solve this issue, then I guess we can close this.

What are your thoughts?

(FYI the error log had no other lines/details)

**Comment by evocateur** [comment]
`lerna publish` and `lerna version` _are_ already decoupled if you always use the [`from-git`](https://github.com/lerna/lerna/tree/master/commands/publish#bump-from-git) or [`from-package`](https://github.com/lerna/lerna/tree/master/commands/publish#bump-from-package) positional argument to `lerna publish`.

```
$ lerna version
$ lerna publish from-git
```

`npm publish` has never cared whether or not a given set of code that it publishes was versioned and tagged in git. Even if `lerna publish` without arguments (implicitly calling `lerna version`) fails during the "upload tarballs to the registry" part, there are more than enough tools in Lerna today to mitigate this occasional failure.

There's no need for "atomic" tag rollbacks or whatever, because the validity of a given commit + tag is not affected by the status of the tarballs present in the configured registry. It sucks when they get out of sync, sure, but that's not a problem Lerna (or any tool) can solve. The problem is "humans," in this instance.

TL;DR: Lerna won't support duplicate version tags, nor will it support publishing duplicate versions of a package to any registry, regardless of its (mis-)capabilities.

**Comment by WoodyWoodsta** [comment]
> TL;DR: Lerna won't support duplicate version tags, nor will it support publishing duplicate versions of a package to any registry, regardless of its (mis-)capabilities.

I understand this and support it. But I think what I am trying to highlight is that currently, Lerna makes the assumption that if a git tag of a particular version exists, then that version definitely exists on the registry (and the important point here is that it makes this assumption in _that_ order - I think this would no longer be an issue if the registry was the point of truth). At the same time, its error handling during `publish` is _brittle_, which introduces a significantly heightened risk of modules going out of sync. It's a command line interface, a _user_ tool. It cannot expect a) network and environment conditions to be stable at all times of operation and b) user's to be 100% correct in their use of the CLI. The fact that there are a number of other, similar issues (https://github.com/lerna/lerna/issues/1917, https://github.com/lerna/lerna/issues/1912, https://github.com/lerna/lerna/issues/455, https://github.com/lerna/lerna/issues/1785, https://github.com/lerna/lerna/issues/1709, https://github.com/lerna/lerna/issues/1668, https://github.com/lerna/lerna/issues/524, https://github.com/lerna/lerna/issues/1578, https://github.com/lerna/lerna/issues/1643, https://github.com/lerna/lerna/issues/1813) illustrates this point.

In any case, there is absolutely no obligation for the tool to respect the patchy, intermittent nature of web technology, and as such, Lerna is not doing anything _wrong_. I think that this is just an area for improvement, in making the experience of using Lerna better, especially for very large projects (high number of modules) as well as those learning how to use it and the registries. Feel free to close if you wish!

**Comment by evocateur** [comment]
@WoodyWoodsta I appreciate your sensitive analysis (and digging through the issues to find those you linked)!

#1767 sketches out a lot of my goals for `lerna publish` in the next major version. One of them is to teach `lerna publish` how to automatically infer `from-git`, `from-package`, and "nightly"/`--canary` publish (which depends on `lerna version` being completely severed).

**Comment by arturi** [comment]
Hi, thanks for Lerna! Hit this as well during publish in Uppy http://github.com/transloadit/uppy:

```
lerna ERR! Error: Command failed: git tag @uppy/companion@0.17.0 -m @uppy/companion@0.17.0
lerna ERR! fatal: tag '@uppy/companion@0.17.0' already exists
```

**Comment by seeliang** [comment]
my workaround is [delete tag](https://nathanhoad.net/how-to-delete-a-remote-git-tag/)

**Comment by AlanSl** [comment]
For anyone who lands on this because a publish run hit an error after the new version git tags were created, but without successfully publishing the packages, and attempting to re-publish after resetting the package.json changes hits this error: 

It looks like it works if you add the `from-package` keyword to your `lerna publish` command.

**Comment by denissabramovs** [comment]
Same here, just got it today. I have to re-publish manually with incremented version to fix inconsistencies between npm registry and git repository tags.

**Comment by denissabramovs** [comment]
It would be nice if we could enforce that, by providing some additional argument, which would override inconsistent version from any side and continue ci/cd successfully.

**Comment by seeliang** [comment]
should be like add git tag after publish ?

**Comment by LucasRuy** [comment]
any news on this topic?

**Comment by JamesHenry** [comment]
Closing in favour of #455
